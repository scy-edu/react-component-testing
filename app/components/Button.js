'use strict'

import React, { Component } from 'react'
import {
  TouchableOpacity,
  Text,
  StyleSheet,
} from 'react-native'
import PropTypes from 'prop-types';

// export styles
export const styles = StyleSheet.create({
  default: {
    backgroundColor: 'papayawhip',
  },
  primary: {
    backgroundColor: 'pink',
  },
});

const Button = ({
  label,
  type,
  onClick,
}) => (
  <TouchableOpacity style={ styles[type] } onPress={ onClick }>
    <Text>
      { label }
    </Text>
  </TouchableOpacity>
);

Button.propTypes = {
  label: PropTypes.string.isRequired,
  type: PropTypes.oneOf(['default', 'primary']).isRequired,
  onClick: PropTypes.func.isRequired,
};

Button.defaultProps = {
  label: 'Submit',
  type: 'default',
  onClick: () => {},
};

Button.displayName = 'Button';

export default Button;
