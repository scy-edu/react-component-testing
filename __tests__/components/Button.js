'use strict';

import 'react-native';
import React from 'react';
import Button, {
  styles,
} from '../../app/components/Button';

import renderer from 'react-test-renderer';
import {
  shallow,
} from 'enzyme';

// create test props
const createTestProps = () => ({
  onClick: () => {},
  label: 'Submit',
  type: 'default',
});

describe('<Button />', () => {

  let wrapper;

  describe('rendering', () => {

    // run before each test
    beforeEach(() => {
      wrapper = shallow(<Button { ...createTestProps() } />);
    });

    it('renders correctly', () => {
      const tree = renderer.create(
        <Button label='Send' />
      );
    });

    it('should render a label', () => {
      expect(wrapper.find('Text').contains('Submit')).toBe(true);
    });

    it('should render a <TouchableOpacity />', () => {
      expect(wrapper.find('TouchableOpacity')).toHaveLength(1);
    });

    describe('no type', () => {
      it('should have the default style', () => {
        expect(wrapper.find('TouchableOpacity').prop('style')).toBe(styles.default);
      });
    });

    describe('primary type', () => {

      // run before each test
      beforeEach(() => {
        wrapper = shallow(<Button
          label='Submit'
          type='primary'
          />);
      });

      it('should have the primary style', () => {
        expect(wrapper.find('TouchableOpacity').prop('style')).toBe(
          styles.primary
        );
      });

    });
  });

  describe('interaction', () => {
    let wrapper;
    let props;

    beforeEach(() => {
      props = {
        label: 'Submit',
        onClick: jest.fn(), // spy: wrapper around normal function to inspect how many times it was called
      };

      wrapper = shallow(<Button { ...props } />);
    });

    describe('clicking the button', () => {
      beforeEach(() => {
        wrapper.find('TouchableOpacity').prop('onPress')();
      });

      it('should call the onClick callback', () => {
        expect(props.onClick).toHaveBeenCalledTimes(1);
      });

    });
  });
});
