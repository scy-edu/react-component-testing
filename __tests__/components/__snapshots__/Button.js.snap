// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> rendering renders correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "papayawhip",
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    Submit
  </Text>
</View>
`;
